kernel/generic/src/mm/as.c

```diff
@@ -2163,175 +2021,65 @@ bool used_space_insert(as_area_t *area, uintptr_t page, size_t count)
-				 * interval of the leaf but its removal
-				 * requires both updating the size of the
-				 * original interval and also inserting a new
-				 * interval.
-				 */
-				new_cnt = ((left_pg + P2SZ(left_cnt)) -
-				    (page + P2SZ(count))) >> PAGE_WIDTH;
-				leaf->value[leaf->keys - 1] -= count + new_cnt;
-				btree_insert(&area->used_space, page +
-				    P2SZ(count), (void *) new_cnt, leaf);
-				goto success;
-			}
-		}
 
+	/* Check for conflict with right interval */
+	if (b != NULL && overlaps(page, P2SZ(count), b->page, P2SZ(b->count)))
 		return false;
-	}
 
-	/*
-	 * The border cases have been already resolved.
-	 * Now the interval can be only between intervals of the leaf.
-	 */
-	btree_key_t i;
-	for (i = 1; i < leaf->keys - 1; i++) {
-		if (page < leaf->key[i]) {
-			uintptr_t left_pg = leaf->key[i - 1];
-			size_t left_cnt = (size_t) leaf->value[i - 1];
-
-			/*
-			 * Now the interval is between intervals corresponding
-			 * to (i - 1) and i.
-			 */
-			if (overlaps(left_pg, P2SZ(left_cnt), page,
-			    P2SZ(count))) {
-				if (page + P2SZ(count) ==
-				    left_pg + P2SZ(left_cnt)) {
-					/*
-					 * The interval is contained in the
-					 * interval (i - 1) of the leaf and can
-					 * be removed by updating the size of
-					 * the bigger interval.
-					 */
-					leaf->value[i - 1] -= count;
-					goto success;
-				} else if (page + P2SZ(count) <
-				    left_pg + P2SZ(left_cnt)) {
-					size_t new_cnt;
-
-					/*
-					 * The interval is contained in the
-					 * interval (i - 1) of the leaf but its
-					 * removal requires both updating the
-					 * size of the original interval and
-					 * also inserting a new interval.
-					 */
-					new_cnt = ((left_pg + P2SZ(left_cnt)) -
-					    (page + P2SZ(count))) >>
-					    PAGE_WIDTH;
-					leaf->value[i - 1] -= count + new_cnt;
-					btree_insert(&area->used_space, page +
-					    P2SZ(count), (void *) new_cnt,
-					    leaf);
-					goto success;
-				}
-			}
-
-			return false;
-		}
+	/* Check if A is adjacent to the new interval */
+	adj_a = (a != NULL) && (a->page + P2SZ(a->count) == page);
+	/* Check if the new interval is adjacent to B*/
+	adj_b = (b != NULL) && page + P2SZ(count) == b->page;
+
+	if (adj_a && adj_b) {
+		/* Fuse into a single interval */
+		a->count += count + b->count;
+		used_space_remove_ival(b);
+	} else if (adj_a) {
+		/* Append to A */
+		a->count += count;
+	} else if (adj_b) {
+		/* Prepend to B */
+		b->page = page;
+		b->count += count;
+	} else {
+		/* Create new interval */
+		ival = slab_alloc(used_space_ival_cache, 0);
+		ival->used_space = &area->used_space;
+		odlink_initialize(&ival->lused_space);
+		ival->page = page;
+		ival->count = count;
+
+		odict_insert(&ival->lused_space, &area->used_space.ivals,
+		    NULL);
 	}
 
-error:
-	panic("Inconsistency detected while removing %zu pages of used "
-	    "space from %p.", count, (void *) page);
-
-success:
-	area->resident -= count;
```

Comment:
Again, this seems to get lost in the new version.

Reply:
OK.